### PR TITLE
Add oauthMetadata and webhookTokenAuthenticators to auth config

### PR DESCRIPTION
**Pull request by enj**
Signed-off-by: Monis Khan <mkhan@redhat.com>

### CODE REVIEW

**Comment by enj** [comment]
@openshift/sig-auth @openshift/api-reviewers

File: config/v1/types_oauth.go [hunk 1]
```diff
@@ -29,11 +29,16 @@ type OAuthSpec struct {
 	// templates allow you to customize pages like the login page.
 	// +optional
 	Templates OAuthTemplates `json:"templates"`
+
+	// masterAuthConfig contains options altering the auth configuration of the kube api server
+	// +optional
+	MasterAuthConfig *MasterAuthConfig `json:"masterAuthConfig"`
```
Comment (deads2k) [inline config/v1/types_oauth.go]: `OAuthSpec` is about how we configure a managed oauth server. This type is about how to configure authentication on the cluster.  I don't think they belong together.  Why wouldn't this be in an authentication type?

File: config/v1/types_oauth.go [hunk 2]
```diff
@@ -542,6 +547,29 @@ type OpenIDClaims struct {
...
+// masterAuthConfig configures authentication options for the kube api server
+type MasterAuthConfig struct {
+	// oauthMetadata contains the discovery endpoint data for OAuth 2.0
+	// Authorization Server Metadata for an external OAuth server.
+	// See IETF Draft:
+	// https://tools.ietf.org/html/draft-ietf-oauth-discovery-04#section-2
+	OAuthMetadata ConfigMapReference `json:"oauthMetadata"`
+
+	// webhookTokenAuthenticators configures remote token reviewers
+	WebhookTokenAuthenticators []WebhookTokenAuthenticator `json:"webhookTokenAuthenticators"`
```
Comment (deads2k) [inline config/v1/types_oauth.go]: I think this piece only belongs in spec.

File: config/v1/types_oauth.go (hunk 1 above)
Comment (enj) [reply config/v1/types_oauth.go]: Yeah I can move both there.

File: config/v1/types_oauth.go (hunk 2 above)
Comment (enj) [reply config/v1/types_oauth.go]: It would set the operator based on the route for osin right?

File: config/v1/types_authentication.go [hunk 3]
```diff
@@ -20,13 +19,27 @@ type Authentication struct {
...
+	// oauthMetadata contains the discovery endpoint data for OAuth 2.0
+	// Authorization Server Metadata for an external OAuth server.
+	// See IETF Draft:
+	// https://tools.ietf.org/html/draft-ietf-oauth-discovery-04#section-2
+	// If oauthMetadata.name is non-empty, this value has precedence
+	// over the observed value stored in status.oauthMetadata
+	// +optional
+	OAuthMetadata ConfigMapReference `json:"oauthMetadata"`
+
+	// webhookTokenAuthenticators configures remote token reviewers
```
Comment (deads2k) [inline config/v1/types_authentication.go]: why would I do this?

File: config/v1/types_authentication.go [hunk 4]
```diff
@@ -20,13 +19,27 @@ type Authentication struct {
 }
 
 type AuthenticationSpec struct {
-	// webhook token auth config (ttl)
-	// external token address
-	// serviceAccountOAuthGrantMethod or remove/disallow it as an option
+	// oauthMetadata contains the discovery endpoint data for OAuth 2.0
+	// Authorization Server Metadata for an external OAuth server.
+	// See IETF Draft:
+	// https://tools.ietf.org/html/draft-ietf-oauth-discovery-04#section-2
```
Comment (deads2k) [inline config/v1/types_authentication.go]: is this just a fancy way of saying, "this is what you serve at .wellknown"?  if so, say that too.

Comment (deads2k) [reply config/v1/types_authentication.go]: This is served from `oc get --raw /asdfklajskdfjasd`

File: config/v1/types_authentication.go [hunk 5]
```diff
@@ -37,3 +50,14 @@ type AuthenticationList struct {
...
+
+// webhookTokenAuthenticator holds the necessary configuration options for a remote token authenticator
+type WebhookTokenAuthenticator struct {
+	// kubeConfig contains kube config file data with the webhook configuration
+	KubeConfig LocalSecretReference `json:"kubeConfig"`
+	// cacheTTL indicates how long an authentication result should be cached.
+	// It takes a valid time duration string (e.g. "5m").
+	// If empty, you get a default timeout of 2 minutes.
+	// If zero (e.g. "0m"), caching is disabled
+	CacheTTL string `json:"cacheTTL"`
```
Comment (deads2k) [inline config/v1/types_authentication.go]: I see no reason to expose this to users.  Please remove.

File: config/v1/types_authentication.go [hunk 6]
```diff
@@ -37,3 +50,14 @@ type AuthenticationList struct {
 	metav1.ListMeta `json:"metadata,omitempty"`
 	Items           []Authentication `json:"items"`
 }
+
+// webhookTokenAuthenticator holds the necessary configuration options for a remote token authenticator
+type WebhookTokenAuthenticator struct {
```
Comment (deads2k) [inline config/v1/types_authentication.go]: leave this struct in case we grow something else.

**Comment by deads2k** [comment]
lgtm otherwise

/approve

**Pull request merged**